### ISSUE TITLE
ssl_generate_random failures after Unity updated to MbedTLS 2.28.7

### ISSUE THREAD

**Issue by MichalPetryka** [Mbed-TLS/mbedtls#9621]
### Summary

Unity Engine (for context, I am not associated with them in any way) uses a statically linked version of MbedTLS for all TLS connection made from both builds and the editor.

After updating to Unity 6000, TLS handshakes stopped working after some, relatively big amount, of sent HTTPS requests and stay broken until the process is restarted.
As suggested by their changelog, they updated to MbedTLS 2.28.7 from 2.28.5 in Unity 6000 (and Unity 2022.3.24+ which also has it is seemingly also affected according to users on the discussion thread):
![image](https://github.com/user-attachments/assets/e4fe8b8f-1385-42d3-80bf-909390fa0eb4)
This is also visible while looking at [PRs merged into their fork](https://github.com/Unity-Technologies/mbedtls/pulls?q=is%3Apr+is%3Aclosed), you can also see they tried to update to 2.28.6 before and decided to go with 2.28.5 instead, which suggests they might have been already experiencing some issues there already.

Debugging the failures by accessing SSL logging (which is not normally exposed by Unity and requires access into their internals) shows that the failing request (and all future ones) error out with (printed with custom formatting):
```
ssl_generate_random() returned -52 (-0x0034)
at C:\build\output\unity\unity\External\Mbedtls\builds\library\ssl_cli.c:L1043
```
which comes from here:
https://github.com/Mbed-TLS/mbedtls/blob/555f84735aecdbd76a566cf087ec8425dfb0c8ab/library/ssl_cli.c#L1043
which in turns comes from here:
https://github.com/Mbed-TLS/mbedtls/blob/555f84735aecdbd76a566cf087ec8425dfb0c8ab/library/ssl_cli.c#L900

This error seems to be this:
https://github.com/Mbed-TLS/mbedtls/blob/555f84735aecdbd76a566cf087ec8425dfb0c8ab/include/mbedtls/ctr_drbg.h#L45

Looking at what Unity does, they seem to not override the rng used by MbedTLS and instead use the default:
https://github.com/Mbed-TLS/mbedtls/blob/555f84735aecdbd76a566cf087ec8425dfb0c8ab/library/ctr_drbg.c#L580
which in turn suggests the error comes likely from:
https://github.com/Mbed-TLS/mbedtls/blob/555f84735aecdbd76a566cf087ec8425dfb0c8ab/library/ctr_drbg.c#L368-L380

`f_entropy` seems to be set to the default `mbedtls_entropy_func`, defined here:
https://github.com/Mbed-TLS/mbedtls/blob/555f84735aecdbd76a566cf087ec8425dfb0c8ab/library/entropy.c#L321
They do not seem to be using `mbedtls_entropy_add_source` anywhere in their code.

As far as I can see, Unity doesn't seem to patch their MbedTLS fork in any way so I assume the same issue is present on normal MbedTLS 2.28.7, I am unable to tell if it also affects MbedTLS 3.x releases though.

### System information

Mbed TLS version (number or commit id): 2.28.7
Operating system and version: at least Windows 10 & 11, not sure if others are affected
Configuration (if not default, please attach `mbedtls_config.h`): not known
Compiler and options (if you used a pre-built binary, please indicate how you obtained it): MbedTLS static linked into Unity
Additional environment information: This was reproduced from Unity 6000.0.3 and 6000.0.17

### Expected behavior

TLS handshakes continue to work

### Actual behavior

TLS handshakes stop working until the process is restarted after a `ssl_generate_random` error.

### Steps to reproduce

Not 100% established but sending multiple concurrent HTTPS requests seems to trigger it after some time.

### Additional information

[Unity discussions thread](https://discussions.unity.com/t/ssl-connection-issues-after-upgrading-to-unity-6-resolved-by-downgrading-to-unity-2021-3-30/1519231) about this issue.

**Comment by davidhorstmann-arm** [comment]
Hi, thanks for the well-investigated report!

I've had a careful look at the changes we made between 2.28.5 and 2.28.7 and I can't see what would be causing this. The 2.28.6 and 2.28.7 releases were very small - one was a license change, the other contained small fixes for RSA and X.509 parsing (double checked with the output of `git diff mbedtls-2.28.5 mbedtls-2.28.7`.

I don't think either of these could have caused a problem with entropy, but maybe there is something I have missed. Is it possible that some other part of the system is causing the entropy sources to fail?

**Comment by MichalPetryka** [comment]
I haven't investigated the diff carefully but Unity `2022.3.23f1`, the latest release before the update to 2.28.7 did not have this issue. The affected project was updated to 6000 after that (we only saw that .24 also changed the MbedTLS version when reporting the issue) and it started seeing the issue, also according to users on the forum, further 2022.3 builds also have the issue there (haven't checked that though).

It could be that Unity changed something with how they interact with MbedTLS when updating to 2.28.7 and that might be causing the issue, it's hard to tell that for certain however due to having no access to Unity sources. Even in such case however, the error log comes from MbedTLS code so unless they have some patch that's not present in their fork, I'd assume it's still likely not Unity specific.

**Comment by gilles-peskine-arm** [comment]
> After updating to Unity 6000, TLS handshakes stopped working after some, relatively big amount, of sent HTTPS requests and stay broken until the process is restarted.

Assuming that Mbed TLS has its default configuration, this is consistent with a delayed failure of entropy gathering. The CTR_DRBG module requests entropy initially, and then it requests entropy again after 10,000 blocks of ≤16 bytes. I don't know exactly how much CTR_DRBG output a TLS connection consumes, and it depends on the cipher suite, but the order of magnitude would be around 10 blocks, so maybe around 100 to 5000 TLS connections.

As a debugging help, if you call `mbedtls_ctr_drbg_set_reseed_interval` with a small value or `mbedtls_ctr_drbg_set_prediction_resistance` with `resistance=1` on the `mbedtls_ctr_drbg_context` object, you can make the error happen sooner.

I can't think of another reason for `mbedtls_ctr_drbg_random` to start failing after it's already succeeded once on a given context, assuming that the context object isn't corrupted. In particular, it doesn't allocate memory. And in any case, I'd expect a different error for a memory allocation failure or a context corruption (but of course you never know with memory corruption).

> they updated to MbedTLS 2.28.7 from 2.28.5

As David noted, there were no changes to entropy or CTR_DRBG in that interval. So I think whatever is going on is not related to the Mbed TLS version change.

(Or maybe some security improvement caused Mbed TLS to consume more DRBG output, and the problem was always present but now comes sooner?)

Can you hook a debugger to the process and find where the error comes from? On Windows, `mbedtls_entropy_func` should eventually call the Windows implementation of `mbedtls_platform_entropy_poll` which calls `CryptGenRandom`. If that call is failing, Mbed TLS can't do anything about it.

**Comment by MichalPetryka** [comment]
I think more info on what we found out when reproducing and investigating will be useful here:

Unity seems to have built an abstraction layer called `UnityTLS` on top of MbedTLS, which they then use for 2 things:

1. They have a custom CURL SSL backend which they then expose using their `UnityWebRequest` class in C#.
2. They also patch the normal C# `HttpClient` and `SslStream` to use `UnityTLS` instead of the default SSL handling.

In the 1st case, we have no access to any MbedTLS nor UnityTLS APIs nor contexts since all the interactions with those are done in a native binary which has everything static linked.
In the 2nd case, we still don't have access to MbedTLS but Unity has internal APIs for UnityTLS accessible from C# and with some patching we enabled logging on every UnityTLS context created there, which also enabled MbedTLS logging.

Interesting thing is, we've been unable to reproduce the error while only using HTTPS through the 2nd path, I've achieved over 150k requests at 30 concurrent ones with no errors. However, when using just the 1st one or both, we do see the issue and it also affects the 2nd path.
It could be that their CURL SSL backend has a bug that corrupts some global context, it could also just be a result of CURL doing some different threading or something similar.

> around 100 to 5000 TLS connections.

A few thousand connections seems to agree with the comments on the Unity forum and our testing.

> As a debugging help, if you call `mbedtls_ctr_drbg_set_reseed_interval` with a small value or `mbedtls_ctr_drbg_set_prediction_resistance` with `resistance=1` on the `mbedtls_ctr_drbg_context` object, you can make the error happen sooner.

As I've mentioned in this comment, we sadly not have access to the MbedTLS contexts nor APIs directly, somebody from Unity would need to do that.

> Can you hook a debugger to the process and find where the error comes from?

I don't think there's any easy way to set a breakpoint in the precompiled binary from Unity that has everything static linked sadly.

> the Windows implementation of `mbedtls_platform_entropy_poll` which calls `CryptGenRandom`. If that call is failing, Mbed TLS can't do anything about it.

I assume that's probably not the case since earlier versions of Unity worked just fine here. Seeing how the issue only manifests with concurrent connections and seemingly only the CURL path is able to cause it, it feels to me like there's some threading issue instead that causes some corruption that leads to entropy failures, can't say whether that would be on the MbedTLS or Unity CURL backend side though (Unity does compile MbedTLS with `MBEDTLS_THREADING_C` enabled btw).

**Comment by gilles-peskine-arm** [comment]
Corruption of the entropy context is indeed a plausible cause of `MBEDTLS_ERR_CTR_DRBG_ENTROPY_SOURCE_FAILED`. When `MBEDTLS_THREADING_C` is enabled, `mbedtls_ctr_drbg_random` should be thread-safe, assuming that nothing else accesses the same `mbedtls_entropy_context`. It would be quite unusual to have something else access the entropy context.

This seems hard to investigate without having debugging symbols in Unity!

Can you at least trace the return values from `CryptGenRandom` or the underlying system call? That would allow you to know whether the cause of the error is external or internal.

**Comment by BioeJD** [comment]
I'm still trying to confirm that my test build did in fact use the appropriate version of MbedTLS, but I ran the following test to help debug this issue:

1. I checked that this issue **didn't** happen to me using Unity 2022.3.23, but **does** happen in 2022.3.24.
2. I pulled the Unity source code for 2022.3.42 (one of the most recent LTS versions where we are seeing this problem).
3. I reverted MbedTLS from 2.28.7 to 2.28.5, which is the version used in 2022.3.23.
4. I compiled this test version of the Unity Editor and Unity Player.
5. I built our test app using my custom Editor version with the downgraded MbedTLS.

We did still see this problem after reverting to 2.28.5. I didn't see other changes in the Unity source code that would relate to TLS/SSL. I also tried swapping out UnityEngine.TLSModule.dll in the test build, and we again still saw the problem.

Again, I'm looking to ensure that there were no flaws in my setup. I'll report back.

**Comment by CloudburstSys** [comment]
[Unity has acknowledged this issue and patched the issue on their side](https://issuetracker.unity3d.com/issues/tlsexception-handshake-failed-error-code-unitytls-internal-error-error-when-a-lot-of-web-requests-are-sent-within-a-few-minutes), considering they say "our MbedTLS builds" I assume they simply forked this project and have made their own changes, which introduced the bug.

Either way, with Unity having fixed the issue for future builds of unity (6000.1.X) with consideration of fixing for past builds (2021.3.X, 2022.3.X, 6000.0.X), I believe this issue can be closed.

**Comment by gilles-peskine-arm** [comment]
Thanks for reporting back! Given that the bug seems to be in how Unity uses or configures Mbed TLS, not in Mbed TLS itself, I am closing this issue.
